### PR TITLE
BUG: Time alignments between RAW and annotate_* funcs

### PR DESCRIPTION
**Pull request by sappelhoff**
intended to fix the failures from https://github.com/mne-tools/mne-python/pull/10067#issuecomment-989791374

The first commit adds a line to a test, which should run ... but which will make the test fail.

The failures most likely have to do with some weird interaction between `raw._first_time` / `raw.first_samp` and the `annotations.orig_time`.

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=17167&view=logs&j=dded70eb-633c-5c42-e995-a7f8d1f99d91&t=8394206f-a00b-5cb7-44ca-464c9d5f9e25&l=3010

I don't know yet how it could be solved, help would be appreciated.

### CODE REVIEW

**Comment by sappelhoff** [comment]
Looking at the ASCII art in the docstring here: https://mne.tools/stable/generated/mne.Annotations.html?highlight=annotations#mne.Annotations

I tried the following approaches, all targeting https://github.com/mne-tools/mne-python/blob/151b8a6cf92524d7e97ea63592b6d3707c38e2b4/mne/preprocessing/artifact_detection.py#L329

1. add `raw._first_time` to the onsets
2. add a `datetime.timedelta(raw._first_time)` to `raw.info["meas_date"]`, i.e., `orig_time = raw.info["meas_date"] + datetime.timedelta(raw._first_time)`

neither worked "out of the box".

**Comment by sappelhoff** [comment]
I pushed a couple more tests, most of which will fail - but should actually pass. I feel like I really opened a can of worms here. So sorry about that ... especially because I still don't know how to fix it.

cc @larsoner @agramfort  @hoechenberger 

I collect loose ideas here: https://github.com/sappelhoff/mne-python/pull/3

**Comment by larsoner** [comment]
@sappelhoff I'll take a look early next week

**Comment by agramfort** [comment]
to make the test pass I need to do:

```
diff --git a/mne/preprocessing/flat.py b/mne/preprocessing/flat.py
index 5aa23647c5..01ab5b72f9 100644
--- a/mne/preprocessing/flat.py
+++ b/mne/preprocessing/flat.py
@@ -98,7 +98,10 @@ def annotate_flat(raw, bad_percent=5., min_duration=0.005, picks=None,
                    (': %s' % (bads,)) if bads else ''))
     bads = [bad for bad in bads if bad not in raw.info['bads']]
     starts, stops = np.array(starts), np.array(stops)
-    onsets = (starts + raw.first_samp) / raw.info['sfreq']
+    if raw.info['meas_date'] is not None:
+        onsets = (starts + raw.first_samp) / raw.info['sfreq']
+    else:
+        onsets = starts / raw.info['sfreq']
     durations = (stops - starts) / raw.info['sfreq']
     annot = Annotations(onsets, durations, ['BAD_flat'] * len(onsets),
                         orig_time=raw.info['meas_date'])
```

basically if there is no meas_date the set_annotations is wrt the first available sample
and it ignores the first_samp... I cannot remember why it was like this though....

**Comment by agramfort** [comment]
@sappelhoff I think I fixed your pb. Let me know if anything is unclear.

**Comment by sappelhoff** [comment]
cool, I just tried this on the original problem with `python -u examples/preprocessing/muscle_detection.py` and that works without warnings now.

> Let me know if anything is unclear.

just not sure why `annotate_nan` and `annotate_breaks` don't need a fix.

**Comment by agramfort** [comment]
@sappelhoff if you can check if annotate_nan and annotate_breaks need fixing too that would help a lot. I will then have a look!

great team effort 💪

File: mne/preprocessing/artifact_detection.py [hunk 1]
```diff
@@ -462,6 +463,11 @@ def annotate_break(raw, events=None,
             logger.info(f'Ignoring annotations with descriptions starting '
                         f'with: {", ".join(ignore)}')
     else:
+        # XXX: I think a bug is hiding here with regard to first_samp
+        #      We should extend the test to work on events and see if
+        #      it breaks here.
+        #      "annotations_from_events" takes a "first_samp" param,
+        #      but this is not being passed here!
```
Comment (sappelhoff) [inline mne/preprocessing/artifact_detection.py:470]: I pushed a commit that shows that `annotate_nan` still fails.

AND I think this :point_up: is a hint that for `annotate_break` we might be missing something as well.

Comment (sappelhoff) [reply mne/preprocessing/artifact_detection.py:470]: update: pushed a commit that reveals that annotate_break is still broken

**Comment by agramfort** [comment]
@sappelhoff I pushed some fixes. The problem is the following. When you set_annotations with annotations that have no meas_date the set_annotations operates taking the first available sample as the 0 of time. In other words set_annotations basically ignores the first_samp when meas_date is None. The consequence of this is that raw.set_annotations(annots) will update the onsets to have raw.annotations now taking into account the first_time. I think the original design decision was that raw.set_annotations should be as easy as possible (ie based on raw.times and hence ignoring first_samp in the absence of meas_date).

does it make sense?

File: mne/preprocessing/artifact_detection.py (hunk 1 above)
Comment (sappelhoff) [reply mne/preprocessing/artifact_detection.py:470]:
```suggestion
```

**Review by agramfort — APPROVED** [review]
@sappelhoff ok for you?

@larsoner green button is yours

**Comment by sappelhoff** [comment]
Yes, thanks Alex. I think we should overhaul the docs from https://mne.tools/stable/generated/mne.Annotations.html?highlight=annotations#mne.Annotations and replace the ASCII art with 4 to six bullet points in english - maybe some time in the future.

**Comment by hoechenberger** [comment]
> I think we should overhaul the docs from https://mne.tools/stable/generated/mne.Annotations.html?highlight=annotations#mne.Annotations and replace the ASCII art with 4 to six bullet points in english - maybe some time in the future.

I think we should include a proper flow chart …

**Comment by agramfort** [comment]
@sappelhoff can you push a tiny addition to the docstring?

@hoechenberger let's focus on more urgent stuff

**Comment by sappelhoff** [comment]
> can you push a tiny addition to the docstring?

I'm out of time :grimacing: need to submit my phd thesis early next year :exploding_head:

File: mne/preprocessing/flat.py [hunk 2]
```diff
@@ -98,7 +98,10 @@ def annotate_flat(raw, bad_percent=5., min_duration=0.005, picks=None,
                    (': %s' % (bads,)) if bads else ''))
     bads = [bad for bad in bads if bad not in raw.info['bads']]
     starts, stops = np.array(starts), np.array(stops)
-    onsets = (starts + raw.first_samp) / raw.info['sfreq']
+    if raw.info['meas_date'] is not None:
+        onsets = (starts + raw.first_samp) / raw.info['sfreq']
+    else:
+        onsets = starts / raw.info['sfreq']
```
Comment (larsoner) [inline mne/preprocessing/flat.py:104]: I don't really like that it's done this way here but differently above. I'll push a commit to unify these

**Review by larsoner — APPROVED** [review]

**Comment by larsoner** [comment]
Good team work! Will merge once CIs come back happy

**Review by hoechenberger — APPROVED** [review]

**Comment by agramfort** [comment]
Thx @sappelhoff !

**Pull request merged**